src/mokernels/independent.jl

```diff
@@ -23,20 +23,39 @@ struct IndependentMOKernel{Tkernel<:Kernel} <: MOKernel
     kernel::Tkernel
 end
 
+# kernel function should be symmetric
+# would really like (κ::IndependentMOKernel)((x, px)::Tuple{T,Int}, (y, py)::Tuple{T,Int}) where T, but seems to cause autodiff problems
 function (κ::IndependentMOKernel)((x, px)::Tuple{Any,Int}, (y, py)::Tuple{Any,Int})
     if px == py
         return κ.kernel(x, y)
     else
-        return 0.0
+        retType = Base.return_types(κ.kernel, (typeof(x), typeof(y)))[1]
+        return zero(retType)
```

Comment:
Using `Base.return_types` always makes me nervous, because there's no guarantee that it will infer the correct type. If possible, it's generally considered best practice to rely on the promotion system in this kind of situation.
```suggestion
        return false
```
be sufficient for your use case? `false` gets promoted to `0f0` when it encounters a `Float32` and `0.0` when it encounters a `Float64`, so maybe that's the best thing to do here?

The whole thing won't be type-stable of course, but it's only a small type-union, so hopefully it's okay

Reply:
In a away, this becomes far less of an issue if you are ok with my proposal for a specific implementation for the kernelmatrix (see kernelmatrix2 and the top level script).

I just think that if possible, type stability is a nice thing to have.

Reply:
Yeah -- I think we'll probably be fine with them. Would you mind posting the timing results so that we don't all have to run them locally?

> I just think that if possible, type stability is a nice thing to have.

Totally agree in general. In this particular case, using `Base.return_types` here means that in order for this function to work at all, `κ.kernel` _must_ infer. If the underlying kernel is type-unstable, and `Base.return_types` returns `Any`, `zero(Any)` will error.

Reply:
Of course, have added the results to the script. 

Ideally, all possible underlying kernels would be type stable and tested as well. But I see how that is a problem to 100% guarantee at this point.

Reply:
Playing around with it some more, and `return false` does not work. 

`k.(x, permutedims(x)` returns a `Matrix{Real}` and multiplying the resulting matrix with a `rand(...)` vector yields a `Vector{Any}`, which I think is not great. 

I think I would rather see if some tests can be written to check type stability for all scalar kernels, but then I am not sure what happens to heterogeneous inputs.

Reply:
Thinking about this a bit more, this suggestion would actually be totally fine for `kernelmatrix_diag`, because `px == py` is always true, and given that you've added `kernelmatrix`, I _think_ this fallback will only ever get hit for `kernelmatrix_diag` in general usage (unless someone is calling the kernel directly, which is generally useful for debugging but isn't what people should be doing if they need performance.

Reply:
Sorry, I missed these comments. I like the suggestion, it also returns the correct result. Have changed it.